frontend/packages/core-auth/src/types/response.ts

```diff
@@ -83,15 +92,30 @@ export class TokenResponse implements TokenResponseConfig {
     return !(TokenResponse.isTokenFresh(this) || !this.refreshToken);
   }
 
+  getExpiryMs(): number {
+    const tenPercentMargin = 0.9;
+    return 1000 * (this.issuedAt + (this.expiresIn ?? 0) * tenPercentMargin);
+  }
+
   async refreshAsync(
     config: Omit<TokenRequestConfig, 'grantType' | 'refreshToken'>,
     discovery: Pick<DiscoveryDocument, 'token_endpoint'>,
-  ) {
-    const request = new RefreshTokenRequest({ ...config, refreshToken: this.refreshToken });
-    const response = await request.performAsync(discovery);
+  ): Promise<this> {
+    const request = new RefreshTokenRequest({
+      ...config,
+      refreshToken: this.refreshToken,
+    });
+    let response: TokenResponse;
+    try {
+      response = await request.performAsync(discovery);
+    } catch (e) {
+      throw new Error(`Error encountered while performing token refresh: ${e}`);
+    }
+
     response.refreshToken = response.refreshToken ?? this.refreshToken;
     const json = response.getRequestConfig();
     this.applyResponseConfig(json);
+
     return this;
```

Comment:
Why not this? The bit after the try/catch seems like it will never run on error and if it did then `response` could be `undefined` and we would error on `response.refreshToken = response.refreshToken ?? this.refreshToken;`. Putting it all in the try prevents this and also allows the response to be a `const` (I typically try to avoid using `let` unless a variable has a strong case for being reassigned).
```
try {
  const response = await request.performAsync(discovery);
  response.refreshToken = response.refreshToken ?? this.refreshToken;
  const json = response.getRequestConfig();
  this.applyResponseConfig(json);
  return this;
} catch (e) {
  throw new Error(`Error encountered while performing token refresh: ${e}`);
}
```